April refactor

AntoineDao opened the pull request:
It's happening! The great Spring cleaning of Queenbee.

This PR aims to fix/resolve the following issues:

- [ ] Function reuse-ability
- [ ] DAG artifact/parameter passing between tasks
- [ ] Nested DAGs and Workflows (like actually testing it out to figure what it would look like)

mostaphaRoudsari commented on queenbee/schema/function.py:
```diff
@@ -22,92 +132,123 @@ class Function(BaseModel):
 
     command: str = Field(
-        ...,
+        '...',
```
This shouldn't be a string. The original one looks correct to me.

AntoineDao replied:
Nope, using `...` without quotes means that this is a mandatory field and has no default value. if we use `"..."` then we are saying that if you do not input a command into your function then the command = ...

mostaphaRoudsari replied:
I see. But why would you want your default command to be `...`?

AntoineDao replied:
I don't. The existing set up just means this field is mandatory.

AntoineDao replied:
Man we were in agreement the whole time. I misread the changes here for ones that you were proposing rather than ones I had made :see_no_evil:

mostaphaRoudsari commented on queenbee/schema/function.py:
```diff
@@ -22,92 +132,123 @@ class Function(BaseModel):
-        None,
-        description='A dictionary of key:values for environmental variables.'
-    )
+    @classmethod
```
This should probably be a `staticmethod` not a `classmethod`.

mostaphaRoudsari commented on queenbee/schema/function.py:
```diff
@@ -22,92 +132,123 @@ class Function(BaseModel):
+        """Validate referenced values"""
+
+        for ref in variables:
+            add_info = ''
+            ref = ref.replace('{{', '').replace('}}', '').strip()
+            add_info = qbvar._validate_inputs_outputs_var_format(ref)
```
We should reuse the methods from `parser` modules or other modules that we have. Using find and replace here and there will make it very hard to debug it in the long term.

AntoineDao replied:
I'm on the fence about this one as it turns out that quite a few of those functions were not re-used in any other schema. ie: there were some parser functions for workflows, some for functions and others for DAGs. As such it felt like it made sense to just stick them in the matching schema classes.

That being said this might change if we move from using `{{...}}` notation to reference objects as we discussed a few days ago. 

Mind waiting a few commits to see what it looks like in a few days?

mostaphaRoudsari commented on queenbee/schema/function.py:
```diff
@@ -22,92 +132,123 @@ class Function(BaseModel):
+            input_names=input_names,
+            variables=ref_var
+        )
+
```
`return v` is missing.

AntoineDao replied:
Ah... that's why it was acting funny :upside_down_face: Good catch thanks!

AntoineDao commented:
Commit 1171778 expects the following two example schemas to be correct:

<details>
  <summary>Operator Example</summary>
```yaml
name: honeybee-radiance

version: 1.2.3

appVersion: 2.3.2

config:
  docker:
    image: ladybugtools/honeybee-radiance:2.3.2
    workdir: /opt/run/

keywords:
- honeybee
- radiance
- ladybug-tools

maintainers:
- name: mostapha
  email: mostapha@ladybug.tools

home: https://ladybug.tools/honeybee-radiance/docs

sources:
- https://github.com/ladybug-tools/honeybee-radiance

icon: https://www.ladybug.tools/assets/img/logo.png

description: An operator to run all things honeybee and radiance!

functions:

- name: 10000-lux-sky
  description: Generates a 10000 lux sky
  command: honeybee radiance sky illuminance 100000
  outputs:
    artifacts:
    - name: sky
      description: The resulting sky object
      path: sky.sky

- name: split-grid
  inputs:
    parameters:
    - name: grid-count
      required: true
    artifacts:
    - name: grid
      description: The input grid to split
      path: grid.pts
  command: honeybee radiance grid split grid.pts --folder output --log-file output/grids.txt
  outputs:
    parameters:
    - name: grid-list
      path: output/grids.txt
    artifacts:
    - name: output-grids-folder
      path: output

- name: create-octree
  description: Generate an octree
  inputs:
    artifacts:
    - name: model
      path: model.json
    - name: sky
      path: sky.sky
  command: honeybee radiance oconv -s sky.sky -m model.json --output scene.oct
  outputs:
    artifacts:
    - name: scene-file
      path: scene.oct

- name: ray-tracing
  description: Run ray tracing using some input data!
  inputs:
    parameters:
    - name: radiance-parameters
      description: a string of radiance parameters
      default: -b 5
    artifacts:
    - name: grid
      path: grid.pts
    - name: scene-file
      path: scene.oct
  command: rtrace -I -h {{inputs.parameters.radiance-parameters}} scene.oct < grid.pts > grid.res
  outputs:
    artifacts:
    - name: result-file
      path: grid.res

- name: post-process
  description: Post process your results and $$$
  inputs:
    artifacts:
    - name: raw-results-folder
      path: raw
  command: honeybee radiance grid merge raw --folder output
  outputs:
    artifacts:
    - name: post-process-folder
      path: output
    parameters:
    - name: daylight-factor-max
      path: output/max
    - name: daylight-factor-average
      path: output/avg
```
</details>


<details>
  <summary>Recipe Example</summary>
```yaml
name: daylight-factor

version: 0.0.1

keywords:
- ladybug-tools
- radiance

maintainers:
- name: mostapha
  email: mostapha@ladybug.tools

icon: https://www.ladybug.tools/assets/img/logo.png

description: An awesome workflow to run daylight factor!

dependencies:

- type: operator
  name: honeybee-radiance
  version: 1.2.3
  owner: ladybug-tools
  source: https://pollination.github.io/garden

flow:
  entrypoint: main
  dags:

  - name: main
    inputs:
      parameters:
      - name: sensor-grid-count
        description: The maximum number of grid points per parallel execution
        default: 100
      - name: radiance-parameters
        description: The radiance parameters for ray tracing
        default: -I -ab 2 -h
      
      artifacts:
      - name: model
        description: A Honeybee model with radiance properties
        required: true
      - name: input-grid
        description: A grid file
        required: true

    tasks:

    - name: generate-sky
      template: honeybee-radiance/10000-lux-sky
      outputs:
        artifacts:
        - name: sky
          path: asset/sky/10000_lux.sky

    - name: split-grid
      template: honeybee-radiance/split-grid
      inputs:
        parameters:
        - name: grid-count
          from:
            type: input
            variable: sensor-grid-count
        artifacts:
        - name: grid
          from:
            type: input
            variable: input-grid
      outputs:
        parameters:
        - name: grid-list
        artifacts:
        - name: output-grids-folder
          path: output/temp

    - name: create-octree
      template: honeybee-radiance/create-octree
      dependencies:
      - generate-sky
      inputs:
        artifacts:
        - name: model
          from:
            type: input
            variable: model
        - name: sky
          from:
            type: tasks
            name: generate-sky
            variable: sky
      outputs:
        artifacts:
        - name: scene-file
          path: output/octree/scene.oct
      
    - name: daylight-factor-simulation
      template: honeybee-radiance/ray-tracing
      dependencies:
      - split-grid
      - create-octree
      loop:
        from:
          type: tasks
          name: split-grid
          variable: grid-list
        sub_folder:
        - item.name
      inputs:
        parameters:
        - name: radiance-parameters
          from:
            type: input
            variable: radiance-parameters
        artifacts:
        - name: grid
          from:
            type: tasks
            name: split-grid
            variable: output-grids-folder
          subpath: {{item.name}}
        - name: scene-files
          from:
            type: tasks
            name: create-octree
            variable: scene-file
      outputs:
        artifacts:
        # This file is persisted in a loop sub-folder so the name won't clash
        - name: results-file
          path: grid.res

    - name: post-process
      template: honeybee-radiance/merge-results
      dependencies:
      - daylight-factor-simulation
      inputs:
        artifacts:
        - name: raw-results-folder
          from:
            type: tasks
            name: daylight-factor-simulation
            variable: results-file
      outputs:
        artifacts:
        - name: post-process-folder
          path: output/post-process
        parameters:
        - name: daylight-factor-average


    outputs:
      artifacts:
      - name: data
        from:
          type: tasks
          name: post-process
          variable: post-process-folder
      parameters:
      - name: average
        from:
          type: tasks
          name: post-process
          variable: daylight-factor-average
```
</details>

The tests are expected to fail. That's for another commit :wink:

AntoineDao commented:
@mostaphaRoudsari sorry I had to overwrite the changes you made as I had made quite a few changes here and there and it was creating conflicts I wasn't brave enough to resolve at 10pm... Sorry about that :disappointed:

AntoineDao commented:
As discussed with @mostaphaRoudsari I am merging the current changes into a Dev branch so both myself and Mostapha can make changes that won't impact master.
